Mathlib/Tactic/NormNum/Factors.lean

```diff
@@ -0,0 +1,156 @@
+    Nat.primeFactorsList n = l :=
+  let ⟨h₁, h₂, h₃⟩ := H Nat.prime_two
+  have := List.chain'_iff_pairwise.1 (@List.Chain'.tail _ _ (_ :: _) h₁)
+  (List.eq_of_perm_of_sorted
+    (Nat.primeFactorsList_unique h₃ h₂) this (Nat.primeFactorsList_sorted _)).symm
+
+open Lean Elab Tactic Qq
+
+/-- Given `n` and `a` (in expressions `en` and `ea`) corresponding to literal numerals
+(in `enl` and `eal`), returns `(l, ⊢ factorsHelper n a l)`. -/
+private partial def evalPrimeFactorsListAux
+    {en enl : Q(ℕ)} {ea eal : Q(ℕ)} (hn : Q(IsNat $en $enl)) (ha : Q(IsNat $ea $eal)) :
+    MetaM ((l : Q(List ℕ)) × Q(FactorsHelper $en $ea $l)) := do
+  /-
+  In this function we will use the convention that `hm_*` are proof about the _meta_ code;
+  these will not actually be used in the construction of the proof, and are simply used to help the
+  reader reason about why the proof construction is correct.
+  -/
+  let n := enl.natLit!
+  let ⟨hm_n0⟩ ← if h : 0 < n then pure <| PLift.up h else
+    throwError m!"{enl} must be positive"
+  let a := eal.natLit!
+  let b := n.minFac
+  let ⟨hm_ab⟩ ← if h : a ≤ b then pure <| PLift.up h else
+    throwError m!"{q($eal < $(enl).minFac)} does not hold"
+  if hm_bn : b < n then
+    -- the factor is less than `n`, so we are not done; remove it to get `m`
+    let m := n / b
+    have em : Q(ℕ) := mkRawNatLit m
+    have hm : Q(IsNat $em $em) := q(⟨rfl⟩)
+    if hm_ba_eq : b = a then
+      -- if the factor is our minimum `a`, then recurse without changing the minimum
+      have h : Q($eal * $em = $en) :=
+        have : a * m = n := by simp [m, ← hm_ba_eq, Nat.mul_div_cancel' (minFac_dvd _)]
+        (q(Eq.refl $en) : Expr)
+      let hp₁ := q(isNat_mul rfl $ha $hm $h)
+      let ⟨el, hp₂⟩ ← evalPrimeFactorsListAux hm ha
+      pure ⟨q($ea :: $el), q(($hp₂).cons_self _ $hp₁)⟩
+    else
+      -- Otherwise when we recurse, we should use `b` as the new minimum factor. Note that
+      -- we must use `evalMinFac.core` to get a proof that `b` is what we computed it as.
+      have eb : Q(ℕ) := mkRawNatLit b
+      have hb : Q(IsNat $eb $eb) := q(⟨rfl⟩)
+      have hbm : Q($eb * $em = $en) :=
+        have : b * m = n := Nat.mul_div_cancel' (minFac_dvd _)
+        (q(Eq.refl $en) : Expr)
+      have hp₁ := q(isNat_mul rfl $hb $hm $hbm)
+      have hp₂ : Q(Nat.blt $ea $eb = true) :=
+        have : a < b := lt_of_le_of_ne hm_ab <| Ne.symm hm_ba_eq
+        (q(Eq.refl (true)) : Expr)
+      let .isNat _ lit hp₃ ← evalMinFac.core q($eb) q(inferInstance) q($eb) hb b | failure
+      assertInstancesCommute
+      have : $lit =Q $eb := ⟨⟩
+      let ⟨l, p₄⟩ ← evalPrimeFactorsListAux hm hb
+      pure ⟨q($eb :: $l), q(($p₄).cons _ $hp₁ $hp₂ $hp₃ )⟩
+  else
+    -- the factor is our number itself, so we are done
+    have hm_bn_eq : b = n := (minFac_le hm_n0).eq_or_lt.resolve_right hm_bn
+    if hm_ba : b = a then
+      have h : Q($en = $ea) :=
+        have : n = a := hm_bn_eq.symm.trans hm_ba
```

Comment:
I like the `hm_*` convention, but it only helps readability if it is used everywhere.

Reply:
btw -- why does the unusedHaves linter not complain?

Reply:
I think that linter only detects tactic-mode haves, not `doElem` haves.